Prevent dll proxing

oid2000 opened the pull request:
First, thanks for your great project!
Could you add an optional call to the SetDllDirectory API with an empty string (i.e. ::SetDllDirectory(_T(""));) in windows EXE loader?
That should prevent DLL proxying attacks on compressed executables in Windows.

Commit:
fix kerning on ascii art logo

Mattiwatti committed:
Make PE load config directory address dword aligned

markus-oberhumer committed:
Post-release version bump.

markus-oberhumer committed:
Prepare for switch to semantic versioning.

markus-oberhumer committed:
Update testsuite checksums caused by version bump.

markus-oberhumer committed:
Use new ASCII art logo.

jreiser committed:
MIPS stub mprotect() suffered from [non-]inlining.

https://github.com/upx/upx/issues/342
https://github.com/upx/upx/issues/339
	modified:   src/stub/src/include/linux.h
   consequences:
	modified:   .github/travis_testsuite_1-expected_sha256sums.sh
	modified:   src/stub/mips.r3000-linux.elf-fold.h
	modified:   src/stub/mipsel.r3000-linux.elf-fold.h
	modified:   src/stub/tmp/mips.r3000-linux.elf-fold.map
	modified:   src/stub/tmp/mipsel.r3000-linux.elf-fold.map

jreiser committed:
ELF shared lib: PT_DYNAMIC might be below xct_off, such as MIPS

https://github.com/upx/upx/issues/298
	modified:   p_lx_elf.cpp

bitraid committed:
Fix unpacking of ASLR enabled PE files

jreiser committed:
Merge pull request #361 from bitraid/fix-pe-aslr-unpack

Fix unpacking of ASLR enabled PE files

bitraid committed:
Don't use --strip-reloc with -d

Always restore the relocation section of Portable Executables on
unpacking, if it exists in the compressed image.

jreiser committed:
More bounds checking for DT_INIT (etc.) in shlib

https://github.com/upx/upx/issues/364
	modified:   p_lx_elf.cpp

jreiser committed:
WIP but should concentrate on section .e_shstrndx instead.

	modified:   p_vmlinx.cpp

jreiser committed:
getElfSections concentrates on _Shdr[.e_shstrndx]

https://github.com/upx/upx/issues/363
	modified:   p_vmlinx.cpp

jreiser committed:
More checking .e_shstrndx

https://github.com/upx/upx/issues/366
	modified:   p_vmlinx.cpp

jreiser committed:
unpack() checks PT_DYNAMIC.p_filesz

https://github.com/upx/upx/issues/367
	modified:   p_lx_elf.cpp

jreiser committed:
check_pt_dynamic() checks PT_DYNAMIC.p_memsz

https://github.com/upx/upx/issues/368
	modified:   p_lx_elf.cpp

jreiser committed:
invert_pt_dynamic() DT_GNU_HASH trims hash array for Rust and Android

https://github.com/upx/upx/issues/369
	modified:   p_lx_elf.cpp

jreiser committed:
--preserve-build-id for ET_EXEC

https://github.com/upx/upx/issues/371
	modified:   p_lx_elf.cpp

jreiser committed:
DT_GNU_HASH check the buckets

https://github.com/upx/upx/issues/365
	modified:   p_lx_elf.cpp

jreiser committed:
max_bucket versus symbias

	modified:   p_lx_elf.cpp

jreiser committed:
ELF layout for page holes when runtime PAGE_SIZE < .p_align

amd64 works
	modified:   p_lx_elf.cpp
	modified:   stub/src/amd64-linux.elf-entry.S
	modified:   stub/src/amd64-linux.elf-fold.S
	modified:   stub/src/amd64-linux.elf-main.c
	modified:   Makefile
	modified:   stub/amd64-linux.elf-entry.h
	modified:   stub/amd64-linux.elf-fold.h
	modified:   stub/amd64-linux.shlib-init.h
	modified:   stub/arm64-linux.elf-fold.h
	modified:   stub/powerpc64-linux.elf-fold.h
	modified:   stub/powerpc64le-linux.elf-fold.h
	modified:   stub/src/amd64-linux.shlib-init.S
	modified:   stub/tmp/amd64-linux.elf-entry.bin.dump
	modified:   stub/tmp/amd64-linux.elf-fold.map
	modified:   stub/tmp/amd64-linux.shlib-init.bin.dump
	modified:   stub/tmp/arm64-linux.elf-fold.map
	modified:   stub/tmp/powerpc64-linux.elf-fold.map
	modified:   stub/tmp/powerpc64le-linux.elf-fold.map

jreiser committed:
unmap C_TEXT

	modified:   stub/src/amd64-linux.elf-fold.S
	modified:   stub/amd64-linux.elf-fold.h
	modified:   stub/tmp/amd64-linux.elf-fold.map

jreiser committed:
un-debug

	modified:   stub/src/amd64-linux.elf-fold.S
	modified:   stub/amd64-linux.elf-fold.h
	modified:   stub/tmp/amd64-linux.elf-fold.map

jreiser committed:
arm64 works: ELF layout for page holes

	modified:   stub/src/arm64-linux.elf-entry.S
	modified:   stub/src/arm64-linux.elf-fold.S

	modified:   stub/arm64-linux.elf-entry.h
	modified:   stub/arm64-linux.elf-fold.h
	modified:   stub/tmp/arm64-linux.elf-entry.bin.dump
	modified:   stub/tmp/arm64-linux.elf-fold.map

jreiser committed:
arm.v4a works: ELF layout for page holes

	modified:   p_lx_elf.cpp
	modified:   stub/src/i386-linux.elf-main.c
	modified:   stub/src/arm.v4a-linux.elf-entry.S
	modified:   stub/src/arm.v4a-linux.elf-fold.S

	modified:   stub/arm.v4a-linux.elf-entry.h
	modified:   stub/arm.v4a-linux.elf-fold.h
	modified:   stub/arm.v5a-linux.elf-entry.h
	modified:   stub/arm.v5a-linux.elf-fold.h
	modified:   stub/armeb.v4a-linux.elf-entry.h
	modified:   stub/armeb.v4a-linux.elf-fold.h
	modified:   stub/i386-linux.elf-fold.h
	modified:   stub/mips.r3000-linux.elf-fold.h
	modified:   stub/mipsel.r3000-linux.elf-fold.h
	modified:   stub/powerpc-linux.elf-fold.h
	modified:   stub/tmp/arm.v4a-linux.elf-entry.bin.dump
	modified:   stub/tmp/arm.v4a-linux.elf-fold.map
	modified:   stub/tmp/arm.v5a-linux.elf-entry.bin.dump
	modified:   stub/tmp/arm.v5a-linux.elf-fold.map
	modified:   stub/tmp/armeb.v4a-linux.elf-entry.bin.dump
	modified:   stub/tmp/armeb.v4a-linux.elf-fold.map
	modified:   stub/tmp/i386-linux.elf-fold.map
	modified:   stub/tmp/mips.r3000-linux.elf-fold.map
	modified:   stub/tmp/mipsel.r3000-linux.elf-fold.map
	modified:   stub/tmp/powerpc-linux.elf-fold.map

jreiser committed:
i386 works: ELF layout for page holes

	modified:   stub/src/i386-linux.elf-entry.S
	modified:   stub/src/i386-linux.elf-fold.S
	modified:   stub/src/i386-linux.elf-main.c

	modified:   stub/arm.v4a-linux.elf-fold.h
	modified:   stub/arm.v5a-linux.elf-fold.h
	modified:   stub/armeb.v4a-linux.elf-fold.h
	modified:   stub/i386-linux.elf-entry.h
	modified:   stub/i386-linux.elf-fold.h
	modified:   stub/mips.r3000-linux.elf-fold.h
	modified:   stub/mipsel.r3000-linux.elf-fold.h
	modified:   stub/powerpc-linux.elf-fold.h
	modified:   stub/src/include/linux.h
	modified:   stub/tmp/arm.v4a-linux.elf-fold.map
	modified:   stub/tmp/arm.v5a-linux.elf-fold.map
	modified:   stub/tmp/armeb.v4a-linux.elf-fold.map
	modified:   stub/tmp/i386-linux.elf-entry.bin.dump
	modified:   stub/tmp/i386-linux.elf-fold.map
	modified:   stub/tmp/mips.r3000-linux.elf-fold.map
	modified:   stub/tmp/mipsel.r3000-linux.elf-fold.map
	modified:   stub/tmp/powerpc-linux.elf-fold.map

jreiser committed:
powerpc32 works: ELF layout for page holes

	modified:   stub/src/powerpc-linux.elf-entry.S
	modified:   stub/src/powerpc-linux.elf-fold.S

	modified:   stub/powerpc-linux.elf-entry.h
	modified:   stub/powerpc-linux.elf-fold.h
	modified:   stub/tmp/powerpc-linux.elf-entry.bin.dump
	modified:   stub/tmp/powerpc-linux.elf-fold.map

jreiser committed:
More checking of DT_GNU_HASH and DT_HASH tables

https://github.com/upx/upx/issues/380
	modified:   p_lx_elf.cpp

jreiser committed:
More checking of DT_GNU_HASH and DT_HASH tables

https://github.com/upx/upx/issues/381
	modified:   p_lx_elf.cpp

jreiser committed:
invert_pt_dynamic() more checking

https://github.com/upx/upx/issues/378
https://github.com/upx/upx/issues/379
	modified:   p_lx_elf.cpp

jreiser committed:
uintptr_t for MSVC

	modified:   p_lx_elf.cpp

markus-oberhumer committed:
Use "python2" instead of "python".

jreiser committed:
powerpc64 works: ELF layout for page holes

	modified:   stub/src/powerpc64le-linux.elf-entry.S
	modified:   stub/src/powerpc64le-linux.elf-fold.S

	modified:   stub/powerpc64-linux.elf-entry.h
	modified:   stub/powerpc64-linux.elf-fold.h
	modified:   stub/powerpc64le-linux.elf-entry.h
	modified:   stub/powerpc64le-linux.elf-fold.h
	modified:   stub/tmp/powerpc64-linux.elf-entry.bin.dump
	modified:   stub/tmp/powerpc64-linux.elf-fold.map
	modified:   stub/tmp/powerpc64le-linux.elf-entry.bin.dump
	modified:   stub/tmp/powerpc64le-linux.elf-fold.map

jreiser committed:
Merge pull request #362 from bitraid/disable-strip-reloc-unpack

Don't use --strip-reloc with -d

jreiser committed:
DPRINTF the munmap

	modified:   stub/src/i386-linux.elf-main.c

jreiser committed:
clarity; no functional change

	modified:   stub/src/powerpc-linux.elf-entry.S

jreiser committed:
phdr[C_BASE].p_memsz for ET_EXEC

	modified:   p_lx_elf.cpp

jreiser committed:
MIPS32 works: ELF layout for page holes

	modified:   stub/src/include/linux.h
	modified:   stub/src/mipsel.r3000-linux.elf-entry.S
	modified:   stub/src/mipsel.r3000-linux.elf-fold.S

	modified:   stub/mips.r3000-linux.elf-entry.h
	modified:   stub/mips.r3000-linux.elf-fold.h
	modified:   stub/mipsel.r3000-linux.elf-entry.h
	modified:   stub/mipsel.r3000-linux.elf-fold.h
	modified:   stub/tmp/mips.r3000-linux.elf-entry.bin.dump
	modified:   stub/tmp/mips.r3000-linux.elf-fold.map
	modified:   stub/tmp/mipsel.r3000-linux.elf-entry.bin.dump
	modified:   stub/tmp/mipsel.r3000-linux.elf-fold.map

jreiser committed:
workaround: MIPS.r3000 early uClibc forgot memset(,0,) of local array

ldso/ldso/ldso.c: _dl_get_ready_to_run()
    _dl_memset(app_tpnt, 0, sizeof(*app_tpnt));
	modified:   stub/src/mipsel.r3000-linux.elf-fold.S

	modified:   stub/mips.r3000-linux.elf-fold.h
	modified:   stub/mipsel.r3000-linux.elf-fold.h

jreiser committed:
use UPX_RSIZE_MAX_MEM

	modified:   p_mach.cpp

jreiser committed:
Merge branch 'devel' of https://github.com/upx/upx into jreiser-pageholes

jreiser committed:
Merge branch 'devel' of https://github.com/upx/upx into devel

jreiser committed:
Mach-o LC_SEGMENTs not monotonic in file (__DWARF for 'go')

https://github.com/upx/upx/issues/222  go-1.14 (but not go-1.14.3 ?)
	modified:   ../p_mach.cpp
	modified:   ../p_mach.h
	modified:   ../p_unix.cpp
	modified:   ../p_unix.h
	modified:   ../stub/src/amd64-darwin.macho-main.c

	modified:   ../stub/amd64-darwin.macho-fold.h
	modified:   ../stub/arm.v5a-darwin.macho-fold.h
	modified:   ../stub/arm64-darwin.macho-fold.h
	modified:   ../stub/tmp/amd64-darwin.macho-fold.map
	modified:   ../stub/tmp/arm.v5a-darwin.macho-fold.map
	modified:   ../stub/tmp/arm64-darwin.macho-fold.map

jreiser committed:
De-compress (upx -d) for ELF pageholes

	modified:   p_lx_elf.cpp

jreiser committed:
Sync testsuite checksums for "ELF pageholes" feature

	modified:   .github/travis_testsuite_1-expected_sha256sums.sh

jreiser committed:
Merge branch 'jreiser-pageholes' into devel

jreiser committed:
Merge branch 'devel' into jreiser-macho2

jreiser committed:
Documentation only

	modified:   stub/src/amd64-darwin.macho-main.c

jreiser committed:
More DPRINTF debugging

	modified:   stub/src/amd64-darwin.macho-main.c

jreiser committed:
Set .b_extra for hdr, especially for Mach-o for 'go'.

https://github.com/upx/upx/issues/222
	modified:   p_unix.cpp

jreiser committed:
DT_HASH chains might be trimmed

https://github.com/upx/upx/issues/383
	modified:   p_lx_elf.cpp

jreiser committed:
DT_HASH chains might be trimmed (64-bit, too)

https://github.com/upx/upx/issues/383
	modified:   p_lx_elf.cpp

jreiser committed:
Unpack: Phdrs must be within expansion of first compressed block

https://github.com/upx/upx/issues/388
	modified:   p_lx_elf.cpp

jreiser committed:
Defend against junk PT_DYNAMIC

https://github.com/upx/upx/issues/390
	modified:   p_lx_elf.cpp

jreiser committed:
Defend against bad PT_DYNAMIC

https://github.com/upx/upx/issues/391
	modified:   p_lx_elf.cpp
	modified:   p_lx_elf.h

jreiser committed:
Avoid 0==nbucket

https://github.com/upx/upx/issues/393
	modified:   p_lx_elf.cpp

jreiser committed:
Check .sh_offset and .sh_size in SHT_DYNAMIC and SHT_STRNDX

https://github.com/upx/upx/issues/394
	modified:   p_lx_elf.cpp

jreiser committed:
Check Shdr more

https://github.com/upx/upx/issues/395
	modified:   p_lx_elf.cpp

jreiser committed:
Check de-compressed SHT_SYMTAB

https://github.com/upx/upx/issues/396
	modified:   p_lx_elf.cpp

jreiser committed:
0==e_shnum obviates testing of e_shoff

	modified:   p_lx_elf.cpp

bitraid committed:
Don't pack Cotrol Flow Guard PE files

jreiser committed:
Merge pull request #399 from bitraid/no-cfg-pe

Don't pack Cotrol Flow Guard PE files

jreiser commented:
First, changes must go into the `devel` branch for testing, not into the `master` branch.  Then, changes to Windows EXE loader of UPX would be changes to files in src/stub/src, such as amd64-win64.pep.S and i386-win32.pe.S.  I don't see either of those files in this pull request. Look at the other `call` instructions, such as to LoadLibraryA and GetProcAddress, for examples to follow.

jreiser committed:
Merge pull request #335 from Mattiwatti/fix-loadconf-align

PE: fix load configuration directory alignment

jreiser committed:
Fix 64-bit ET_EXEC

https://github.com/upx/upx/issues/405
	modified:   p_lx_elf.cpp
	modified:   stub/src/amd64-linux.elf-main.c

	modified:   ../.github/travis_testsuite_1-expected_sha256sums.sh
	modified:   stub/amd64-linux.elf-fold.h
	modified:   stub/arm64-linux.elf-fold.h
	modified:   stub/powerpc64-linux.elf-fold.h
	modified:   stub/powerpc64le-linux.elf-fold.h
	modified:   stub/tmp/amd64-linux.elf-fold.map
	modified:   stub/tmp/arm64-linux.elf-fold.map
	modified:   stub/tmp/powerpc64-linux.elf-fold.map
	modified:   stub/tmp/powerpc64le-linux.elf-fold.map

jreiser committed:
MSVC strikes again

	modified:   p_lx_elf.cpp

Mattiwatti committed:
PE: only add a TLS relocation if tls_handler_offset_reloc != 0

Brambor commented on BUGS:
```diff
@@ -1,10 +1,10 @@
-                  888       8   888            .8PY888.
-                  `88.    .8'   888           d8'  `888b
-                    `YbodP'    o888o        o888o  o88888o
+                 ooooo     ooo  ooooooooo.  ooooooo  ooooo
+                 `888'     `8'  `888   `Y88. `8888    d8'
+                  888       8    888   .d88'   Y888..8P
+                  888       8    888ooo88P'     `8888'
+                  888       8    888           .8PY888.
+                  `88.    .8'    888          d8'  `888b
+                    `YbodP'     o888o       o888o  o88888o
```
I don't get it, should this be fixed by rebase?

jreiser committed:
Beware overlap when rebuildImports

Based on a hint from a little bird.
	modified:   pefile.cpp

jreiser committed:
Avoid overlap when rebuild relocs.

Based on a hint from a little bird.
	modified:   packer.cpp

bitraid committed:
Only pack w/ -f PE files that force integrity check

jreiser committed:
fillPackHeader more checks

https://github.com/upx/upx/issues/414
	modified:   packhead.cpp

liamjm committed:
tweaks: some extra checks on lengths

liamjm committed:
tweaks: some extra checks on lengths

liamjm committed:
fix build

liamjm committed:
rebase

jreiser committed:
omemmove exactly as much

Fix 14992260c60b8d6677a677a9cdfae98b11353df7; revealed by testsuite
	modified:   pefile.cpp

jreiser committed:
fillPackHeader: revise and add yet more checks

https://github.com/upx/upx/issues/414
https://github.com/upx/upx/pull/415  revised
	modified:   packhead.cpp

jreiser committed:
Merge pull request #413 from bitraid/pe-integrity

Require -f to pack a PE file that demands run-time integrity checking

jreiser committed:
Merge pull request #407 from Mattiwatti/ignore-bogus-tls-relocs

PE: only add a TLS relocation if tls_handler_offset_reloc != 0

jreiser committed:
Merge pull request #416 from liamjm/packheader_more_checks

tweaks: some extra checks on lengths

jreiser committed:
headway vs size

https://github.com/upx/upx/issues/414
	modified:   packhead.cpp

jreiser commented:
Merging to `master` branch is highly restricted.  If you still want to proceed, then merge to `devel` branch first.
